google/cloud/storage/internal/bucket_metadata_parser.cc

```diff
@@ -67,298 +68,444 @@ StatusOr<UniformBucketLevelAccess> ParseUniformBucketLevelAccess(
+  if (!effective_time) return std::move(effective_time).status();
+  retention_policy = BucketRetentionPolicy{
+      std::chrono::seconds(*retention_period), *effective_time, *is_locked};
+  return Status{};
+}
 
-  if (json.count("website") != 0) {
-    auto website = json["website"];
-    BucketWebsite w;
-    w.main_page_suffix = website.value("mainPageSuffix", "");
-    w.not_found_page = website.value("notFoundPage", "");
-    result.website_ = std::move(w);
-  }
-  return result;
+Status ParseVersioning(absl::optional<BucketVersioning>& versioning,
+                       nlohmann::json const& json) {
+  if (!json.contains("versioning")) return Status{};
+  auto v = json["versioning"];
+  if (!v.contains("enabled")) return Status{};
+  auto enabled = internal::ParseBoolField(v, "enabled");
+  if (!enabled) return std::move(enabled).status();
+  versioning = BucketVersioning{*enabled};
+  return Status{};
 }
 
-StatusOr<BucketMetadata> BucketMetadataParser::FromString(
-    std::string const& payload) {
-  auto json = nlohmann::json::parse(payload, nullptr, false);
-  return FromJson(json);
+Status ParseWebsite(absl::optional<BucketWebsite>& website,
+                    nlohmann::json const& json) {
+  if (!json.contains("website")) return Status{};
+  auto w = json["website"];
+  BucketWebsite value;
+  value.main_page_suffix = w.value("mainPageSuffix", "");
+  value.not_found_page = w.value("notFoundPage", "");
+  website = std::move(value);
+  return Status{};
 }
 
-std::string BucketMetadataToJsonString(BucketMetadata const& meta) {
-  nlohmann::json metadata_as_json;
-  if (!meta.acl().empty()) {
-    for (BucketAccessControl const& a : meta.acl()) {
-      nlohmann::json entry;
-      SetIfNotEmpty(entry, "entity", a.entity());
-      SetIfNotEmpty(entry, "role", a.role());
-      metadata_as_json["acl"].emplace_back(std::move(entry));
-    }
+void ToJsonAcl(nlohmann::json& json, BucketMetadata const& meta) {
+  if (meta.acl().empty()) return;
+  nlohmann::json value;
+  for (BucketAccessControl const& a : meta.acl()) {
+    nlohmann::json entry;
+    SetIfNotEmpty(entry, "entity", a.entity());
+    SetIfNotEmpty(entry, "role", a.role());
+    value.push_back(std::move(entry));
   }
+  json["acl"] = std::move(value);
+}
 
-  if (!meta.cors().empty()) {
-    for (CorsEntry const& v : meta.cors()) {
-      nlohmann::json cors_as_json;
-      if (v.max_age_seconds.has_value()) {
-        cors_as_json["maxAgeSeconds"] = *v.max_age_seconds;
-      }
-      if (!v.method.empty()) {
-        cors_as_json["method"] = v.method;
-      }
-      if (!v.origin.empty()) {
-        cors_as_json["origin"] = v.origin;
-      }
-      if (!v.response_header.empty()) {
-        cors_as_json["responseHeader"] = v.response_header;
-      }
-      metadata_as_json["cors"].emplace_back(std::move(cors_as_json));
+void ToJsonCors(nlohmann::json& json, BucketMetadata const& meta) {
+  if (meta.cors().empty()) return;
+
```

Comment:
nit: most others don't have this newline. change if you want.

Reply:
Done.